File: ai_config_tool/push.py
"""Push local AI IDE configuration back to the canonical repository."""
from __future__ import annotations

import logging
import shutil
import tempfile
from pathlib import Path
from typing import Iterable, List

from .constants import (
    DEFAULT_BRANCH,
    DEFAULT_REMOTE_DIR,
    DEFAULT_REPO_URL,
    SUPPORTED_DIRECTORIES,
)
from .git_utils import clone_repo, run_git
from .path_utils import normalize_remote_dir

LOGGER = logging.getLogger(__name__)


def _existing_local_dirs(target: Path) -> List[Path]:
    present: List[Path] = []
    for directory in SUPPORTED_DIRECTORIES:
        path = target / directory
        if path.exists():
            present.append(path)
    return present


def _copy_directories(
    source_root: Path, dest_root: Path, directories: Iterable[str]
) -> None:
    if dest_root.exists():
        shutil.rmtree(dest_root)
    dest_root.mkdir(parents=True, exist_ok=True)
    for directory in directories:
        src = source_root / directory
        dest = dest_root / directory
        if src.exists():
            dest.parent.mkdir(parents=True, exist_ok=True)
            shutil.copytree(src, dest)


def run_push(
    target: Path,
    repo_url: str = DEFAULT_REPO_URL,
    branch: str = DEFAULT_BRANCH,
    remote_dir: str = DEFAULT_REMOTE_DIR,
    commit_message: str = "chore: sync ai IDE config",
) -> None:
    target = target.expanduser().resolve()
    if not target.exists():
        raise FileNotFoundError(f"Target path {target} does not exist")
    local_dirs = _existing_local_dirs(target)
    if not local_dirs:
        raise RuntimeError(
            "No local .cursor/.claude directories found. Nothing to push."
        )
    remote_dir_path = normalize_remote_dir(remote_dir)
    with tempfile.TemporaryDirectory(prefix="ai-config-push-") as tmpdir:
        repo_path = Path(tmpdir) / "repo"
        LOGGER.info(
            "Cloning %s (branch %s) to prepare push into %s",
            repo_url,
            branch,
            remote_dir_path,
        )
        clone_repo(
            repo_url=repo_url,
            destination=repo_path,
            branch=branch,
            depth=None,
        )
        remote_root = repo_path / remote_dir_path
        if remote_dir_path == Path("."):
            for directory in SUPPORTED_DIRECTORIES:
                dest = repo_path / directory
                if dest.exists():
                    shutil.rmtree(dest)
                src = target / directory
                if src.exists():
                    dest.parent.mkdir(parents=True, exist_ok=True)
                    shutil.copytree(src, dest)
        else:
            _copy_directories(target, remote_root, SUPPORTED_DIRECTORIES)
        status = run_git(["git", "status", "--porcelain"], cwd=repo_path)
        if not status.stdout.strip():
            LOGGER.info("Remote repository already matches local configuration.")
            return
        if remote_dir_path == Path("."):
            for directory in SUPPORTED_DIRECTORIES:
                run_git(["git", "add", directory], cwd=repo_path)
        else:
            run_git(["git", "add", str(remote_dir_path)], cwd=repo_path)
        run_git(["git", "commit", "-m", commit_message], cwd=repo_path)
        LOGGER.info("Pushing changes to origin/%s", branch)
        run_git(["git", "push", "origin", branch], cwd=repo_path)
        LOGGER.info("Push completed successfully.")

File: src/push.js
const { promises: fs } = require('fs');
const { join, resolve } = require('path');
const { tmpdir } = require('os');
const { mkdtemp } = require('fs/promises');
const simpleGit = require('simple-git');

const { DEFAULT_BRANCH, DEFAULT_REMOTE_DIR, DEFAULT_REPO_URL, SUPPORTED_DIRECTORIES } = require('./constants');
const { cloneRepo } = require('./git-utils');
const { normalizeRemoteDir } = require('./utils');

/**
 * Get existing local directories
 */
async function getExistingLocalDirs(target) {
  const existing = [];

  for (const directory of SUPPORTED_DIRECTORIES) {
    const path = join(target, directory);
    try {
      await fs.access(path);
      existing.push(path);
    } catch {
      // Directory doesn't exist
    }
  }

  return existing;
}

/**
 * Copy directories to remote location
 */
async function copyDirectories(
  sourceRoot,
  destRoot,
  directories
) {
  // Remove existing destination root
  try {
    await fs.access(destRoot);
    await fs.rm(destRoot, { recursive: true, force: true });
  } catch {
    // Destination doesn't exist
  }

  // Create destination root
  await fs.mkdir(destRoot, { recursive: true });

  for (const directory of directories) {
    const src = join(sourceRoot, directory);
    const dest = join(destRoot, directory);

    try {
      await fs.access(src);
      await copyDirectoryRecursive(src, dest);
    } catch {
      // Source directory doesn't exist, skip
    }
  }
}

/**
 * Copy directory recursively
 */
async function copyDirectoryRecursive(src, dest) {
  await fs.mkdir(dest, { recursive: true });

  const entries = await fs.readdir(src, { withFileTypes: true });

  for (const entry of entries) {
    const srcPath = join(src, entry.name);
    const destPath = join(dest, entry.name);

    if (entry.isDirectory()) {
      await copyDirectoryRecursive(srcPath, destPath);
    } else {
      await fs.copyFile(srcPath, destPath);
    }
  }
}

/**
 * Main push function
 */
exports.runPush = async function runPush(options) {
  const {
    target = process.cwd(),
    repoUrl = DEFAULT_REPO_URL,
    branch = DEFAULT_BRANCH,
    remoteDir = DEFAULT_REMOTE_DIR,
    commitMessage = 'chore: sync ai IDE config'
  } = options;

  const targetPath = resolve(target);

  // Check if target exists
  try {
    await fs.access(targetPath);
  } catch {
    throw new Error(`Target path ${targetPath} does not exist`);
  }

  // Check if there are local directories to push
  const localDirs = await getExistingLocalDirs(targetPath);
  if (localDirs.length === 0) {
    throw new Error('No local .cursor/.claude/.ai/.trae/.specify directories found. Nothing to push.');
  }

  const remoteDirPath = normalizeRemoteDir(remoteDir);

  // Create temporary directory
  const tempDir = await mkdtemp(join(tmpdir(), 'ai-config-push-'));

  try {
    console.log(`Cloning ${repoUrl} (branch ${branch}) to prepare push into ${remoteDirPath}`);

    // Clone the repository
    const repoPath = join(tempDir, 'repo');
    await cloneRepo({
      repoUrl,
      destination: repoPath,
      branch,
      depth: undefined // Full clone for pushing
    });

    const remoteRoot = join(repoPath, remoteDirPath);
    const git = simpleGit(repoPath);

    if (remoteDirPath === '.') {
      // If remote dir is root, copy directories directly to repo root
      for (const directory of SUPPORTED_DIRECTORIES) {
        const dest = join(repoPath, directory);

        // Remove existing directory
        try {
          await fs.access(dest);
          await fs.rm(dest, { recursive: true, force: true });
        } catch {
          // Directory doesn't exist
        }

        const src = join(targetPath, directory);
        try {
          await fs.access(src);
          await copyDirectoryRecursive(src, dest);
        } catch {
          // Source directory doesn't exist, skip
        }
      }
    } else {
      // Copy directories to remote subdirectory
      await copyDirectories(targetPath, remoteRoot, SUPPORTED_DIRECTORIES);
    }

    // Check if there are any changes
    const status = await git.status();

    let hasChanges = false;
    if (remoteDirPath === '.') {
      // Check only the supported directories for changes
      for (const directory of SUPPORTED_DIRECTORIES) {
        const dirFiles = status.files.filter(file => file.path.startsWith(directory + '/') || file.path === directory);
        if (dirFiles.length > 0) {
          hasChanges = true;
          break;
        }
      }
    } else {
      // Check if remote directory has changes
      const remoteFiles = status.files.filter(file => file.path.startsWith(remoteDirPath + '/'));
      hasChanges = remoteFiles.length > 0;
    }

    if (!hasChanges) {
      console.log('Remote repository already matches local configuration.');
      return;
    }

    // Add files to git
    if (remoteDirPath === '.') {
      // Add each supported directory separately
      for (const directory of SUPPORTED_DIRECTORIES) {
        try {
          await fs.access(join(repoPath, directory));
          await git.add(directory);
        } catch {
          // Directory doesn't exist, skip
        }
      }
    } else {
      // Add the entire remote directory
      await git.add(remoteDirPath);
    }

    // Commit changes
    await git.commit(commitMessage);

    // Push changes
    console.log(`Pushing changes to origin/${branch}`);
    await git.push('origin', branch);

    console.log('Push completed successfully.');

  } finally {
    // Clean up temporary directory
    try {
      await fs.rm(tempDir, { recursive: true, force: true });
    } catch (error) {
      console.warn(`Failed to clean up temporary directory: ${error}`);
    }
  }
};